File: 0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    val: int
    next: "ListNode"

    def __init__(self, val=0, next: "ListNode" = None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # 空间换时间
        if not head:
            return head
        nodes = [head]
        current = head
        while current.next:
            current = current.next
            nodes.append(current)

        length = len(nodes)

        if length == 1:
            return None

        delete_index = length - n

        if delete_index == 0:
            head = head.next
        elif delete_index == length - 1:
            nodes[-2].next = None
        else:
            nodes[delete_index - 1].next = nodes[delete_index].next

        return head

File: 0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // 双指针， 两个指针间隔 n， 将后一个指针指向尾元素，另一个指针则指向被删除元素，为了更方便可以设置 dummy 节点
        ListNode* dummy = new ListNode(0, head);
        ListNode * first = head, *second = dummy;
        for (int i = 0; i < n; i++) {
            first = first->next;
        }
        while (first != nullptr) {
            first = first->next;
            second = second->next;
        }
        second->next = second->next->next;
        ListNode* ans = dummy->next;
        delete dummy;
        return ans;
    }
};
